runtime/compiler/optimizer/HCRGuardAnalysis.cpp

```diff
@@ -88,80 +88,99 @@ bool TR_HCRGuardAnalysis::shouldSkipBlock(TR::Block *block)
-            _regularKillSetInfo[currentBlock->getNumber()]->setAll(getNumberOfBits());
-            _exceptionKillSetInfo[currentBlock->getNumber()]->setAll(getNumberOfBits());
+            _regularKillSetInfo[blockId]->setAll(numBits);
+            _exceptionKillSetInfo[blockId]->setAll(numBits);
             treeTop = currentBlock->getExit();
             }
          continue;
          }
 
-      if (treeTop->getNode()->getOpCode().canRaiseException())
-         {
-         exceptingTTSeen = true;
-         }
-
-      osrNode = NULL;
-      if (comp()->isPotentialOSRPoint(treeTop->getNode(), &osrNode) && !checklist.contains(osrNode))
+      if (comp()->isPotentialOSRPoint(ttNode, &osrNode) && !checklist.contains(osrNode))
          {
          checklist.add(osrNode);
          bool supportsOSR = comp()->isPotentialOSRPointWithSupport(treeTop);
          if (supportsOSR)
             {
-            _regularKillSetInfo[currentBlock->getNumber()]->setAll(getNumberOfBits());
-            if (exceptingTTSeen)
-               _exceptionKillSetInfo[currentBlock->getNumber()]->setAll(getNumberOfBits());
+            _regularKillSetInfo[blockId]->setAll(numBits);
+            _regularGenSetInfo[blockId]->resetAll(numBits);
             }
-         /*else
-            {
-            TR::Node *node = treeTop->getNode();
-            traceMsg(comp(), "mismatch at n%dn %d:%d %s\n", node->getGlobalIndex(), node->getByteCodeInfo().getCallerIndex(), node->getByteCodeInfo().getByteCodeIndex(),  comp()->signature());
-            }*/
-         if (!supportsOSR && treeTop->getNode()->getByteCodeInfo().getCallerIndex() > -1)
+         else
             {
-            _regularGenSetInfo[currentBlock->getNumber()]->setAll(getNumberOfBits());
-            if (exceptingTTSeen)
-               *(_exceptionGenSetInfo[currentBlock->getNumber()]) |= *(_regularGenSetInfo[currentBlock->getNumber()]);
+            _regularGenSetInfo[blockId]->setAll(numBits);
+            _regularKillSetInfo[blockId]->resetAll(numBits);
             }
          }
-      else if (treeTop->getNode()->isTheVirtualGuardForAGuardedInlinedCall() && comp()->cg()->supportsMergingGuards())
+      else if (ttNode->isTheVirtualGuardForAGuardedInlinedCall() && comp()->cg()->supportsMergingGuards())
          {
-         TR_VirtualGuard *guardInfo = comp()->findVirtualGuardInfo(treeTop->getNode());
+         TR_VirtualGuard *guardInfo = comp()->findVirtualGuardInfo(ttNode);
          if (guardInfo->getKind() != TR_HCRGuard)
             {
-            _regularKillSetInfo[currentBlock->getNumber()]->setAll(getNumberOfBits());
-            if (exceptingTTSeen)
-               _exceptionKillSetInfo[currentBlock->getNumber()]->setAll(getNumberOfBits());
+            _regularKillSetInfo[blockId]->setAll(numBits);
+            _regularGenSetInfo[blockId]->resetAll(numBits);
             }
          }
+
+      bool canNodeRaiseException = false;
+      if (ttNode->getOpCode().canRaiseException())
```

Comment:
When the current tree could both yield and throw, I think exception edges should always need to gen, even at OSR points where the analysis kills. It's possible for the tree to yield, allowing assumptions to be invalidated, and then throw afterward. An OSR guard (if one is necessary) would only run after non-exceptional completion of the tree, so it wouldn't stop control from reaching the exception handler.

That said, I'm not sure we need to do anything at all for the exception edges; I thought we had an OSR point at the start of each catch block for this very reason.